### PR TITLE
add utility function for retrieving a trace id

### PR DESCRIPTION
**Pull request by garypen**
add utility function for retrieving a trace id

Retrieve a unique trace id which can be used to link together processing
steps in a plugin (or anywhere else in the router).

If an execution trace context cannot be established, the return value
will be `None` To correctly confirm that two things are linked you
would have to check:

```
let other_id = apollo_router::tracer::maybe_trace();
let my_id = apollo_router::tracer::maybe_trace();

// Note: enough to check that at least one of your traces is some...
if my_id.is_some() && my_id == other_id {
    <blah...>
}
```
Without that first check, i.e.:

```
if my_id == other_id {
    <blah...>
}
```

you could be linking together two unrelated things.

fixes: #1536

### CODE REVIEW

File: apollo-router/src/lib.rs [hunk 1]
```diff
@@ -69,6 +69,7 @@ pub mod services;
 mod spec;
 mod state_machine;
 mod test_harness;
+pub mod tracer;
```
Comment (BrynCooke) [inline apollo-router/src/lib.rs:72]: Would prefer `tracing` as I could see us adding other trace related things in future.

**Review by BrynCooke — APPROVED** [review]
Approved, but see comment on module name.

**Comment by garypen** [comment]
> > If an execution trace context cannot be established, the return value
> > will be TraceId::INVALID (all zeros)
> 
> why not return an `Option`?

I had that first, but decided to mirror the behaviour in the existing interface (which does this). I do think the nice thing about the `Option` is that we'd unlock all the combinators and it would mean we could cut out the check to make sure it's not INVALID.

I'll have a look at re-shaping this with an Option and see if it looks better.

File: apollo-router/src/lib.rs (hunk 1 above)
Comment (garypen) [reply apollo-router/src/lib.rs:72]: I'll change it.

**Review by o0Ignition0o — APPROVED** [review]

**Pull request merged**